### PR TITLE
client: include details about GOAWAYs in status messages

### PR DESCRIPTION
**Pull request by apolcyn**
Based on https://github.com/grpc/grpc-go/pull/4311 - only the last two commits here are new (this is also motivated to debug b/182572215)

Fixes https://github.com/grpc/grpc-go/issues/4266

### CODE REVIEW

File: internal/transport/http2_client.go [hunk 1]
```diff
@@ -1215,12 +1225,18 @@ func (t *http2Client) setGoAwayReason(f *http2.GoAwayFrame) {
 			t.goAwayReason = GoAwayTooManyPings
 		}
 	}
+	var m bytes.Buffer
+	m.WriteString("code: ")
+	m.WriteString(f.ErrCode.String())
+	m.WriteString(", debug data: ")
+	m.Write(f.DebugData())
+	t.goAwayDebugMessage = m.String()
```
Comment (dfawley) [inline internal/transport/http2_client.go:1233]: Is this better than `t.goAwayDebugMessage = fmt.Sprintf("code: %s, debug data: %v", f.ErrCode, f.DebugData)` ?

File: test/end2end_test.go [hunk 2]
```diff
@@ -1335,48 +1337,142 @@ func testConcurrentServerStopAndGoAway(t *testing.T, e env) {
...
+	te.srv.Stop()
+	if _, err := stream.Recv(); err == nil || (!strings.Contains(err.Error(), possibleConnResetMsg) && !strings.Contains(err.Error(), possibleEOFMsg)) {
+		t.Fatalf("%v.Recv() = _, %v, want _, rpc error containing substring: |%v| OR |%v|", stream, err, possibleConnResetMsg, possibleEOFMsg)
 	}
-	if err := stream.Send(sreq); err != nil {
-		t.Fatalf("%v.Send(%v) = %v, want <nil>", stream, sreq, err)
+	close(rpcDoneOnClient)
+}
+
+func (s) TestDetailedGoawayErrorOnGracefulClosePropagatesToRPCError(t *testing.T) {
```
Comment (dfawley) [inline test/end2end_test.go:1378]: Same comment as in the other PR about using StubServer and not enumerating the test environments.

File: test/end2end_test.go [hunk 3]
```diff
@@ -1335,48 +1337,142 @@ func testConcurrentServerStopAndGoAway(t *testing.T, e env) {
...
+
+	cc := te.clientConn()
+	tc := testpb.NewTestServiceClient(cc)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	stream, err := tc.FullDuplexCall(ctx)
+	if err != nil {
+		t.Fatalf("%v.FullDuplexCall = _, %v, want _, <nil>", tc, err)
+	}
+	expectedErrorMessageSubstring := "received prior goaway: code: NO_ERROR"
```
Comment (dfawley) [inline test/end2end_test.go:1415]: `const` please

**Review by dfawley — COMMENTED** [review]
Please also update your branch and resolve conflicts - thanks!

File: test/end2end_test.go (hunk 3 above)
Comment (apolcyn) [reply test/end2end_test.go]: done

File: test/end2end_test.go (hunk 2 above)
Comment (apolcyn) [reply test/end2end_test.go:1385]: done

File: internal/transport/http2_client.go (hunk 1 above)
Comment (apolcyn) [reply internal/transport/http2_client.go]: That looks better, changed to that

**Comment by apolcyn** [comment]
BTW, my last changes also address review comment: https://github.com/grpc/grpc-go/pull/4311#discussion_r612598779, which I forgot to address in #4311

**Comment by apolcyn** [comment]
BTW there's one flaked test here but it seems unrelated to this change:

```
        --- FAIL: Test/FaultInjection_Unary/delay_80%,_abort_50% (8.01s)
            --- FAIL: Test/FaultInjection_Unary/delay_80%,_abort_50%/50%_delay_and_abort (2.00s)
                fault_test.go:528: RPC 0: err: rpc error: code = DeadlineExceeded desc = context deadline exceeded, intnCalls: [], newTimerCalls: []
                fault_test.go:530: WANTED code: Unimplemented, intnCalls: [100 100], newTimerCalls: [3s]
                fault_test.go:528: RPC 1: err: rpc error: code = DeadlineExceeded desc = context deadline exceeded, intnCalls: [], newTimerCalls: []
                fault_test.go:530: WANTED code: Unimplemented, intnCalls: [100 100], newTimerCalls: [3s]
```

File: test/end2end_test.go [hunk 4]
```diff
@@ -1380,6 +1382,84 @@ func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
...
+	rpcDoneOnClient := make(chan struct{})
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			<-rpcDoneOnClient
+			return status.Error(codes.Internal, "arbitrary status")
+		},
+	}
+	sopts := []grpc.ServerOption{
+		grpc.KeepaliveEnforcementPolicy(keepalive.EnforcementPolicy{
+			MinTime: time.Second * 1000, /* arbitrary, large value */
```
Comment (dfawley) [inline test/end2end_test.go:1434]: Fun fact: you can tell a google home "go back / skip ahead a {year,century,millenium(!)}" while playing a song/video to go to the beginning/end.

**Review by dfawley — APPROVED** [review]
Looks great, sorry for the delay!

**Comment by dfawley** [comment]
Could you please update your branch so the new presubmit tests we have will run?

**Comment by apolcyn** [comment]
> Could you please update your branch so the new presubmit tests we have will run?

Done. New tests triggered

**Pull request merged**